fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexMaintainer.java

```diff
@@ -0,0 +1,244 @@
+import org.apache.lucene.document.Document;
+import org.apache.lucene.document.Field;
+import org.apache.lucene.document.SortedDocValuesField;
+import org.apache.lucene.document.StoredField;
+import org.apache.lucene.document.TextField;
+import org.apache.lucene.index.IndexWriter;
+import org.apache.lucene.queryparser.classic.MultiFieldQueryParser;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.util.BytesRef;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+
+import static com.apple.foundationdb.record.lucene.IndexWriterCommitCheckAsync.getOrCreateIndexWriter;
+
+/**
+ * Index maintainer for Lucene Indexes backed by FDB.  The insert, update, and delete functionality
+ * coupled with the scan functionality is implemented here.
+ *
+ */
+@API(API.Status.EXPERIMENTAL)
+public class LuceneIndexMaintainer extends StandardIndexMaintainer {
+    private static final Logger LOG = LoggerFactory.getLogger(LuceneIndexMaintainer.class);
+    private final Analyzer analyzer;
+    protected static final String PRIMARY_KEY_FIELD_NAME = "p"; // Need to find reserved names..
+    private static final String PRIMARY_KEY_SEARCH_NAME = "s"; // Need to find reserved names..
+    private final List<String> fieldNames;
+    private final Executor executor;
+
+    public LuceneIndexMaintainer(@Nonnull final IndexMaintainerState state, @Nonnull Executor executor, @Nonnull Analyzer analyzer) {
+        super(state);
+        this.analyzer = analyzer;
+        this.fieldNames = new ArrayList<>(2);
+        this.state.index.getRootExpression()
+                .normalizeKeyForPositions().forEach( (fke) -> this.fieldNames.add( ((FieldKeyExpression) fke).getFieldName()));
+        this.executor = executor;
+    }
+
+    /**
+     * The scan uses the low element in the range to execute the
+     * MultiFieldQueryParser.
+     *
+     * @param scanType the {@link IndexScanType type} of scan to perform
+     * @param range the range to scan
+     * @param continuation any continuation from a previous scan invocation
+     * @param scanProperties skip, limit and other properties of the scan
+     * @return RecordCursor
+     */
+    @Nonnull
+    @Override
+    public RecordCursor<IndexEntry> scan(@Nonnull IndexScanType scanType, @Nonnull TupleRange range,
+                                         @Nullable byte[] continuation,
+                                         @Nonnull ScanProperties scanProperties) {
+        LOG.trace("scan scanType={}", scanType);
+        Verify.verify(range.getLow() != null);
+        try {
+            final MultiFieldQueryParser parser = new MultiFieldQueryParser(fieldNames.toArray(new String[0]) , analyzer);
+            Query query = parser.parse(range.getLow().getString(0));
```

Comment:
Not for now and really a core issue, but seems like we're missing an abstraction in the index API. Like maybe there are index-supported table functions like there are record functions and aggregate functions. And you use one of those with a query argument rather than a scan.

Reply:
Not for now.